packages/svelte/src/compiler/phases/scope.js

```diff
@@ -100,6 +104,261 @@ export class Binding {
 	}
 }
 
+class Evaluation {
+	/** @type {Set<any>} */
+	values = new Set();
+
+	/**
+	 * True if there is exactly one possible value
+	 * @readonly
+	 * @type {boolean}
+	 */
+	is_known = true;
+
+	/**
+	 * True if the value is known to not be null/undefined
+	 * @readonly
+	 * @type {boolean}
+	 */
+	is_defined = true;
+
+	/**
+	 * True if the value is known to be a string
+	 * @readonly
+	 * @type {boolean}
+	 */
+	is_string = true;
+
+	/**
+	 * True if the value is known to be a number
+	 * @readonly
+	 * @type {boolean}
+	 */
+	is_number = true;
+
+	/**
+	 * @readonly
+	 * @type {any}
+	 */
+	value = undefined;
+
+	/**
+	 *
+	 * @param {Scope} scope
+	 * @param {Expression} expression
+	 */
+	constructor(scope, expression) {
+		switch (expression.type) {
+			case 'Literal':
+				this.values.add(expression.value);
+				break;
+
+			case 'Identifier':
+				var binding = scope.get(expression.name);
+
+				if (binding) {
+					if (
+						binding.initial?.type === 'CallExpression' &&
+						get_rune(binding.initial, scope) === '$props.id'
+					) {
+						this.values.add(STRING);
+						break;
+					}
+
+					const is_prop =
+						binding.kind === 'prop' ||
+						binding.kind === 'rest_prop' ||
+						binding.kind === 'bindable_prop';
+
+					if (!binding.updated && binding.initial !== null && !is_prop) {
+						const evaluation = binding.scope.evaluate(/** @type {Expression} */ (binding.initial));
+						for (const value of evaluation.values) {
+							this.values.add(value);
+						}
+						break;
+					}
+
+					// TODO each index is always defined
+				}
+
+				// TODO glean what we can from reassignments
+				// TODO one day, expose props and imports somehow
+
+				this.values.add(UNKNOWN);
+				break;
+
+			case 'BinaryExpression':
+				var a = scope.evaluate(/** @type {Expression} */ (expression.left)); // `left` cannot be `PrivateIdentifier` unless operator is `in`
+				var b = scope.evaluate(expression.right);
```

Comment:
I see this switch statement is using `var` in a couple places - on purpose or muscle memory from the runtime? I'd rather use `let` etc here, also to make the code below (which just does `a = ...` in some places, which looks like reassignment but actually isn't) easier to reason about.

Reply:
you can't use `let` here without getting the following eslint error...

> Unexpected lexical declaration in case block.eslint ([no-case-declarations](https://eslint.org/docs/latest/rules/no-case-declarations))

...unless you wrap each `case` in a block. Which might be neater actually, though I seem to recall there's some reason you get yelled at for that as well. Will try it